File: build_knowledge_base.py
import json
import os
import re
import time
from pathlib import Path
from typing import Any

from openai import OpenAI


BASE_URL = "https://api.deepseek.com"
MODEL = "deepseek-chat"
CHUNK_CHARS = 18000
MAX_RETRIES = 5


def get_client() -> OpenAI:
    api_key = os.getenv("DEEPSEEK_API_KEY")
    if not api_key:
        raise RuntimeError("请先设置环境变量 DEEPSEEK_API_KEY。")
    return OpenAI(api_key=api_key, base_url=BASE_URL)


def read_txt(filepath: str) -> str:
    path = Path(filepath).expanduser().resolve()
    if not path.exists():
        raise FileNotFoundError(f"文件不存在: {path}")

    for encoding in ("utf-8-sig", "utf-8", "gb18030"):
        try:
            return path.read_text(encoding=encoding)
        except UnicodeDecodeError:
            continue
    raise UnicodeDecodeError("unknown", b"", 0, 1, "无法识别文本编码，请转换为 UTF-8。")


def safe_title(filepath: str) -> str:
    title = Path(filepath).stem
    return re.sub(r'[\\/:*?"<>|\s]+', "_", title).strip("_") or "novel"


def chunk_text(text: str, chunk_chars: int = CHUNK_CHARS) -> list[str]:
    paragraphs = re.split(r"(\n\s*\n)", text)
    chunks: list[str] = []
    current = ""

    for part in paragraphs:
        if len(current) + len(part) <= chunk_chars:
            current += part
            continue
        if current.strip():
            chunks.append(current.strip())
        current = part

    if current.strip():
        chunks.append(current.strip())
    return chunks


def call_llm(client: OpenAI, messages: list[dict[str, str]], json_mode: bool = False) -> Any:
    last_error: Exception | None = None
    for attempt in range(1, MAX_RETRIES + 1):
        try:
            kwargs: dict[str, Any] = {
                "model": MODEL,
                "messages": messages,
                "temperature": 0.2,
            }
            if json_mode:
                kwargs["response_format"] = {"type": "json_object"}
            response = client.chat.completions.create(**kwargs)
            content = response.choices[0].message.content or ""
            return json.loads(content) if json_mode else content
        except Exception as exc:
            last_error = exc
            if attempt == MAX_RETRIES:
                break
            time.sleep(min(2 ** attempt, 30))
    raise RuntimeError(f"LLM 调用失败: {last_error}")


def analyze_chunk(client: OpenAI, chunk: str, index: int, total: int) -> dict[str, Any]:
    system_prompt = (
        "你是中文长篇小说的文学分析师。请只做结构化分析。"
        "可以识别性爱场景的类型和功能，但不要生成具体描写。"
        "输出必须是 JSON 对象，包含 characters、plot_points、relationship_changes、"
        "tone、open_threads、chapter_candidates 字段。"
    )
    user_prompt = (
        f"这是全文分块 {index}/{total}。请用 JSON 分析这一块的角色、事件、关系变化、"
        "情绪基调、悬念和可能章节节点。保持客观，避免补写剧情。\n\n"
        f"{chunk}"
    )
    return call_llm(
        client,
        [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt},
        ],
        json_mode=True,
    )


def merge_analyses(client: OpenAI, title: str, analyses: list[dict[str, Any]]) -> dict[str, Any]:
    system_prompt = (
        "你是长篇小说制片统筹。请把分块分析合并为可供改写系统使用的知识库。"
        "输出必须是 JSON 对象，包含 title、characters、chapter_blueprint、"
        "relationship_map、global_tone、continuity_notes 字段。"
    )
    user_prompt = (
        "请合并下面的分块分析为一个完整 JSON 知识库。要求去重角色、统一称谓、保留主线。\n\n"
        + json.dumps({"title": title, "chunk_analyses": analyses}, ensure_ascii=False)
    )
    return call_llm(
        client,
        [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt},
        ],
        json_mode=True,
    )


def analyze_global_novel(full_text: str, novel_title: str) -> dict[str, Any]:
    client = get_client()
    chunks = chunk_text(full_text)
    analyses: list[dict[str, Any]] = []

    for index, chunk in enumerate(chunks, start=1):
        print(f"分析分块 {index}/{len(chunks)}...")
        analyses.append(analyze_chunk(client, chunk, index, len(chunks)))

    return merge_analyses(client, novel_title, analyses)


def save_knowledge_base(novel_title: str, data: dict[str, Any]) -> Path:
    output_path = Path(f"knowledge_base_{novel_title}.json")
    output_path.write_text(json.dumps(data, ensure_ascii=False, indent=2), encoding="utf-8")
    return output_path


def main() -> None:
    filepath = input("请输入小说 TXT 路径: ").strip().strip('"')
    novel_title = safe_title(filepath)
    full_text = read_txt(filepath)
    data = analyze_global_novel(full_text, novel_title)
    output_path = save_knowledge_base(novel_title, data)
    print(f"知识库已生成: {output_path}")


if __name__ == "__main__":
    main()

File: build_knowledge_base.js
import fs from "node:fs/promises";
import path from "node:path";
import readline from "node:readline/promises";
import { stdin as input, stdout as output } from "node:process";
import OpenAI from "openai";

const BASE_URL = "https://api.deepseek.com";
const MODEL = "deepseek-chat";
const CHUNK_CHARS = 18000;
const MAX_RETRIES = 5;

function getClient() {
  const apiKey = process.env.DEEPSEEK_API_KEY;
  if (!apiKey) {
    throw new Error("请先设置环境变量 DEEPSEEK_API_KEY。");
  }
  return new OpenAI({ apiKey, baseURL: BASE_URL });
}

async function readTxt(filepath) {
  const absolutePath = path.resolve(filepath);
  try {
    return await fs.readFile(absolutePath, "utf8");
  } catch (error) {
    throw new Error(`读取失败，请确认文件为 UTF-8 编码: ${absolutePath}\n${error.message}`);
  }
}

function safeTitle(filepath) {
  const title = path.basename(filepath, path.extname(filepath));
  return title.replace(/[\\/:*?"<>|\s]+/g, "_").replace(/^_+|_+$/g, "") || "novel";
}

function chunkText(text, chunkChars = CHUNK_CHARS) {
  const parts = text.split(/(\n\s*\n)/);
  const chunks = [];
  let current = "";

  for (const part of parts) {
    if (current.length + part.length <= chunkChars) {
      current += part;
      continue;
    }
    if (current.trim()) chunks.push(current.trim());
    current = part;
  }

  if (current.trim()) chunks.push(current.trim());
  return chunks;
}

async function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function callLlm(client, messages, { jsonMode = false, temperature = 0.2 } = {}) {
  let lastError = null;
  for (let attempt = 1; attempt <= MAX_RETRIES; attempt += 1) {
    try {
      const request = {
        model: MODEL,
        messages,
        temperature
      };
      if (jsonMode) request.response_format = { type: "json_object" };
      const response = await client.chat.completions.create(request);
      const content = response.choices[0]?.message?.content ?? "";
      return jsonMode ? JSON.parse(content) : content;
    } catch (error) {
      lastError = error;
      if (attempt === MAX_RETRIES) break;
      await sleep(Math.min(2 ** attempt, 30) * 1000);
    }
  }
  throw new Error(`LLM 调用失败: ${lastError?.message ?? lastError}`);
}

async function analyzeChunk(client, chunk, index, total) {
  const systemPrompt =
    "你是中文长篇小说的文学分析师。请只做结构化分析，不生成露骨色情内容。" +
    "输出必须是 JSON 对象，包含 characters、plot_points、relationship_changes、tone、open_threads、chapter_candidates 字段。";
  const userPrompt =
    `这是全文分块 ${index}/${total}。请用 JSON 分析这一块的角色、事件、关系变化、情绪基调、悬念和可能章节节点。保持客观，避免补写剧情。\n\n${chunk}`;

  return callLlm(
    client,
    [
      { role: "system", content: systemPrompt },
      { role: "user", content: userPrompt }
    ],
    { jsonMode: true }
  );
}

async function mergeAnalyses(client, title, analyses) {
  const systemPrompt =
    "你是长篇小说制片统筹。请把分块分析合并为可供改写系统使用的知识库。" +
    "输出必须是 JSON 对象，包含 title、characters、chapter_blueprint、relationship_map、global_tone、continuity_notes、safety_notes 字段。";
  const userPrompt =
    "请合并下面的分块分析为一个完整 JSON 知识库。要求去重角色、统一称谓、保留主线，不要生成露骨色情内容。\n\n" +
    JSON.stringify({ title, chunk_analyses: analyses });

  return callLlm(
    client,
    [
      { role: "system", content: systemPrompt },
      { role: "user", content: userPrompt }
    ],
    { jsonMode: true }
  );
}

async function analyzeGlobalNovel(fullText, novelTitle) {
  const client = getClient();
  const chunks = chunkText(fullText);
  const analyses = [];

  for (let index = 0; index < chunks.length; index += 1) {
    console.log(`分析分块 ${index + 1}/${chunks.length}...`);
    analyses.push(await analyzeChunk(client, chunks[index], index + 1, chunks.length));
  }

  return mergeAnalyses(client, novelTitle, analyses);
}

async function saveKnowledgeBase(novelTitle, data) {
  const outputPath = `knowledge_base_${novelTitle}.json`;
  await fs.writeFile(outputPath, JSON.stringify(data, null, 2), "utf8");
  return outputPath;
}

async function main() {
  const rl = readline.createInterface({ input, output });
  const filepath = (await rl.question("请输入小说 TXT 路径: ")).trim().replace(/^"|"$/g, "");
  rl.close();

  const novelTitle = safeTitle(filepath);
  const fullText = await readTxt(filepath);
  const data = await analyzeGlobalNovel(fullText, novelTitle);
  const outputPath = await saveKnowledgeBase(novelTitle, data);
  console.log(`知识库已生成: ${outputPath}`);
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
